### ISSUE TITLE
NUnit Framework no longer buildable using VS2015

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#2620]
I decided to do a quick check of an issue by writing a test (#2619 if it matters) so I pulled down master, created a branch and opened VS2015.

Turns out none of the projects load any longer. They do load in VS2017 (with errors in my case) and in researching I discovered the `BUILDING.md` file now says VS2017 is required, where previously we were able to build the framework using VS2012 and higher.

I feel this is a real loss and - to be quite frank - displays something of a disdain for those who don't upgrade to the "latest and greatest" when told to do so by the vendor. With some degree of caring for such folks, this could easily have been done in a less painful way or possibly not done at all.

However, if it had to be done, it should not have been slip-streamed but rather given serious visibility. Only those who participated in the details of the change would have a way to know the consequence. There are people who build NUnit from source and every commit to master is a potential source release for them.

If this change can't be mitigated - well, actually won't is the operative word, since it obviously could be done - then we need to have a pretty big notification going out to those users. IMO, it should be accompanied with an apology for springing this on them.

For me personally, this was a rather unpleasant surprise and appears to remove my ability to contribute to our primary project unless I convert to VS2017. 😢

**Comment by jnm2** [comment]
I thought this was both discussed and decided clearly, so that no one watching the PR should have been surprised: https://github.com/nunit/nunit/pull/2380#issuecomment-339831284

**Comment by jnm2** [comment]
We can remove this line though since I got Mono 5.2 building on Linux:

> Linux support will be limited to .NET Core until Mono catches up

**Comment by CharliePoole** [comment]
So, it only matters that folks watching the PR should know what's coming? I for one was not watching, although I made some comments via my cellphone. My own view was that I didn't like the change but felt it was within the rights of the ongoing team to do it, so long as notice was given to users. The latter is what didn't happen as far as I can see.

My own assumption was that this was going to happen at a release point with a breaking change notice, not simply appear in master, so I was pretty dismayed to run into it. However, it only impacts my ability to contribute, which is a smaller problem than how it impacts users who build from source using VS2012 through VS2015.

**Comment by jnm2** [comment]
No, I'm not addressing your main point. Only that I'm surprised that you were surprised since Rob laid out what he wanted to do pretty clearly and you responded to that. And the objections you're bringing up, that this is disdain for users, were thoroughly discussed and put in balance against other needs. There was quite a bit of outside contributor pain the longer we waited.

About building from source– maybe I should start an issue on that and assign it to myself. I think it should be possible without any version of Visual Studio on the machine.

**Comment by rprouse** [comment]
This was discussed at length in #2380, so I'm sorry that you didn't notice the new requirement. At some point, we need to ask contributors to use newer IDE's or frameworks so that we can support newer frameworks. .NET Standard 2.0 and .NET Core newer than 1.1 does not build in older IDEs. If we want to support them, we need to require VS 2017.

It is unfortunate, but I don't have time to maintain multiple solutions to support older versions of Visual Studio. The new `csproj` format is a big change for Visual Studio, but I think it is a great change that makes it much easier to work with.

Also, I am not sure that our solutions did build in versions of Visual Studio older than 2015 because of some of the newer language features we use and the .NET Core projects. I guess people could have used the Linux solution to get around the . NET Core stuff, but not the language features.

**Comment by rprouse** [comment]
> About building from source– maybe I should start an issue on that and assign it to myself. I think it should be possible without any version of Visual Studio on the machine.

The newer version of MSBuild is likely required, but I think that is downloadable now as a package isn't it?

**Comment by jnm2** [comment]
@rprouse Last time I checked, MSBuild 15 is bin-deployable and has a NuGet package but it doesn't come with any MSBuild SDKs (network of .targets and .props). .NET Core's SDK only includes the right MSBuild SDKs for compiling against net40 and higher. Both Mono and Visual Studio include the right MSBuild SDKs for net20 and higher.

**Comment by CharliePoole** [comment]
@jnm2 If you look back, you'll see I wasn't one of those involved in all that discussion. Two reasons for that. First, it became clear early on that my view was a minority view and it wasn't worth spending too much time arguing for it. Second, I was traveling and limited to a cell phone for commentary.

Frankly, I'm quite amazed that you characterize my responses as positive. I thought I was pretty negative, perhaps even to the point of being uncivil, although I was trying to avoid that. It is hard to find the balance at times!

I'm not surprised that the change is happening, only that it slipped in so quickly, without any announcement, polling of users, etc. It's as if it didn't impact anyone but us.

@rprouse All your reasons for wanting to do this made sense to me - enough sense that I stifled my objections. But slipping it in without discussion at the user level was a big surprise to me. I thought you made it clear that was not going to happen.

You say in your last comment "At some point, we need to ask contributors to use newer IDE's or frameworks..." My point is, we still haven't asked anyone anything, but are simply requiring it. I don't consider the PR discussion to be the place that anything gets properly aired to users or even to the majority of contributors for that matter.

Bottom line, I knew where you were going but I didn't know you were getting there so abruptly.

**Comment by jnm2** [comment]
@CharliePoole Not all your responses 😄 but specifically your response to Rob laying out all the details of the change and what to expect:

> Thanks, that's what I was after. I think this kind of thing eventually belongs in the docs. The vision statement may also be out of sync, but it's written pretty generally so maybe it will do.

Hopefully it's not too amazing for me to read that as positive.

**Comment by CharliePoole** [comment]
Sorry to confuse you. That's me being civil regarding a change that I absolutely hate but which I see as inevitable. I actually thought that was pretty clear to everyone.

I'm down on myself for being unclear and for assuming that notice would be given in advance to those affected. I can only assume (again) that we all had different definitions of "notice" and "those affected."

**Comment by CharliePoole** [comment]
BTW, if what Rob had laid out had actually been published, we wouldn't be having this discussion. I think you are overestimating the visibility of comments in a PR.

**Comment by jnm2** [comment]
So, how do we handle internal codebase announcements at NUnit– say, if we were to allow a different C# version? Just rely on the mailing lists? Announcement issues on GitHub?

**Comment by CharliePoole** [comment]
I can only give you my opinion - how you will handle it in future is up to you. 

In the past, any user concern went to the nunit-discuss list or it's predecessor. Dev concerns went to the dev list. Of course the mail list isn't all our users but we always felt justified in saying that was where they could get the info. Following all our issues and PRs is more than we should expect users to do.

Internal info for contributors is wherever we decide it should be. That has been changing. I think major changes should be announced on the dev list.

If course your example is one where we would be lessening rather than increasing a restriction, which is not so big a problem.

**Comment by rprouse** [comment]
@CharliePoole what would you like us to do with this issue? We cannot support newer targets unless we use VS 2017 and those newer targets are important to me and to a large percentage of our users. It would be nice if we could be inclusive and build on older IDE's, but we can't. To me, asking that users update their build tools if they want to contribute is no different than asking that they have the required versions of .NET installed. The tools are free and easily installed.

**Comment by CharliePoole** [comment]
I decided to wait and cool off before responding. Here's what I think is needed...

1. Users in general need to be told about it. I think the nunit-discuss list is the correct place, but if you feel that there is a broader way to reach people, then that should be used.

2. Users who build from the release source won't be affected until we do a new release. They should be warned about what will happen and if possible how to deal with it.

3. Users who build from the source of master may already have been affected. I think they need an apology, followed by the same info that those building from release source need.

4. At the same time, it may be possible to gain some insight into the numbers of users who build from either source. It's theoretically possible to say we don't support them any longer, but only provide binaries, but I feel that would be contrary to our stated values and those of the open source community.

The most affected users are those who are building from source for legacy platforms. By definition, they are not using newer platforms. It would be entirely possible to give them solution and project files that build in earlier versions of VS if the team feels the effort is justified. IMO, this could wait until learning how many folks are affected. This, by the way, is the step that I understood, from our discussion, was going to be taken ahead of making the change.

On the personal side, I've managed to get VS2017 working for the framework on my system, which id didn't do before. It does seem that the new project format has solved some issues for me. It's rather tedious working due to the inherent slowdown from building six times the number of projects for every rebuild, however. Over the years, I've learned to work and compile in smaller and smaller steps, so I'm having to go back to the old style of less frequent compiles.

If we are able to agree at all on what's to be said, I'm willing to create something for users.

**Comment by jnm2** [comment]
Just a thought. Users that hit breaking changes in our API aren't afraid to voice their expectations loud and clear. That's a good thing. For users who build from source, changing the project format is a much bigger change than an API break. So if there are users who specifically build from master, and they weren't happy with the new format, I'd expect to be hearing from them any time.

**Comment by rprouse** [comment]
> I feel that would be contrary to our stated values and **those of the open source community**

If being able to build from source on legacy platforms was part of the .NET opensource values, all the other opensource projects didn't read the manifesto. **All of the top 15** active, non-Microsoft projects on NuGet, **have switched to the new Visual Studio 2017 format** and require VS2017. Most don't even support compiling on all platforms. NUnit does support compiling **all of our targets** on Windows, Mac and Linux. That was not the case before the switch to the new project format. If anything, we are now more inclusive.

When we switched to requiring Visual Studio 2015 Update 3 (for C# language level 6 and for .NET Standard) we didn't announce it on the mailing list. We only discussed that change in a pull request and an issue. We have also had seven releases since we made that change and nobody has complained that they could not compile from source on their platform of choice.

I don't see why requiring VS2017 is any different than requiring VS2015, but if you feel that it deserves an announcement then we will make one.

**Comment by CharliePoole** [comment]
@rprouse You asked me what I thought was needed. Since I was feeling very angry, I waited until I could write something concilatory. We can see where that got me from your response.

You quoted me entirely out of context. Here's the full quotation...

> 4. At the same time, it may be possible to gain some insight into the numbers of users who build from either source. It's theoretically possible to say we don't support them any longer, but only provide binaries, but I feel that would be contrary to our stated values and those of the open source community.

In other words, the statement had nothing to do with what IDEs are supported but with support for users who build from source. As far as I know, you have not yet proposed that we stop supporting such users, have you? My hope was that point 4, like the other points, was something we actually __agreed__ on but you have managed to find something to dislike about it anyway. That's unfortunate.

Regarding past changes, I'm well aware that we didn't consult users when imposing VS2015 Update 3. I didn't like that either, but I kept relatively quiet about it. This one seemed much more out of line with our published values in requiring use of a specific Vendor's IDE and I couldn't keep quiet and hold my head up.

Your comment about so-called ".NET opensource values" is also out of line. The open source community existed a long time before .NET and will exist after it. I'm well aware that the values I have argued for and which we voted to follow are __not__ those of the .NET community. NUnit has __always__ been exceptional in upholding those values. That's why we take the trouble to put some of them down in writing!

Clearly we are talking about how to deal with a change that is already made, not about reversing it. If by making an announcement you mean simply telling folks what has happened and that they are stuck with it, then I can't see much good coming from that. At a minimum, it seems to me you could ask users if there is anything that can be done to ease the transition for them. What, for example, should a company that builds only for .NET 4.5 do?
